### PR TITLE
rfc: Preload Realm for Service Workers

### PR DESCRIPTION
**Pull request by samuelmaddock**
> Preload realms provide an isolated context to run JavaScript prior to code evaluating in another
context. This expands functionality of Electron's preload scripts to provide new context targets.
This proposal intends to target Service Worker contexts with potential plans for new contexts (web
workers, shared workers) in future proposals.

[Read rendered document](https://github.com/samuelmaddock/rfcs/blob/preload-realm-rfc/text/0002-preload-realm.md)

Recommended prerequisite readings:
- [Design of V8 bindings](https://chromium.googlesource.com/chromium/src/third_party/+/master/blink/renderer/bindings/core/v8/V8BindingDesign.md)
- [ShadowRealms explainer](https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md)

If you're curious, there's a very rough wip prototype being written in https://github.com/samuelmaddock/electron/tree/feat/preload-realm

### CODE REVIEW

**Comment by alexi** [comment]
Also wondering about an update on this. We need modern chrome extensions support in electron, and we currently rely on the electron-chrome-extensions package.

File: text/0002-preload-realm.md [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+Preload realms provide an isolated context to run JavaScript prior to code evaluating in another
+context. This expands functionality of Electron's preload scripts to provide new context targets.
+This proposal intends to target Service Worker contexts with potential plans for new contexts (web
+workers, shared workers) in future proposals.
+
+## Motivation
+
+[Electron's Chrome extension support](https://www.electronjs.org/docs/latest/api/extensions) adds a
+subset of APIs to support the use case of DevTools-related extensions. Extending this to support the
+full API would not only add a large maintenence burdon, but might impose and restrict design of
```
Comment (Kilian) [inline text/0002-preload-realm.md:18]: maintenance burden

File: text/0002-preload-realm.md [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+    script: "sw-preload.js",
+  },
+]);
+```
+
+### `contextBridge`
+
+A service worker's preload script should have access to the initiator context from the preload realm
+context. The goal of bridging objects between two contexts remains the same, but may require new
+terminologies. A "main world" might not make sense for workers so I've elected to use "initiator
```
Comment (Kilian) [inline text/0002-preload-realm.md:239]: I much prefer 'initiator' over 'main' in the current context as well (with the node part often also being called main) so I'd be in favor of this terminology change in general.

**Comment by RitvarsZ** [comment]
Any updates on this? This is a showstopper for our project right now, would love to see this implemented.

**Comment by samuelmaddock** [comment]
GitHub is stuck on "Processing updates", attempting to close and reopen to fix this.

**Comment by samuelmaddock** [comment]
welp

Screenshot 2024-08-06 at 10 19 03 AM

**Pull request closed without merging**